src/test/webServer.ts

```diff
@@ -0,0 +1,140 @@
+    const webServer = new WebServer(config);
+    if (webServer._port)
+      await webServer._verifyFreePort();
+    try {
+      await webServer._startWebServer();
+      await webServer._waitForAvailability();
+    } catch (error) {
+      await webServer.kill();
+      throw error;
+    }
+    process.env.PLAYWRIGHT_WEB_SERVER_URL = `http://localhost:${webServer._port}`;
+    return webServer;
+  }
+
+  private async _verifyFreePort() {
+    assert(this._port);
+    const cancellationToken = { canceled: false };
+    const portIsUsed = await Promise.race([
+      new Promise(resolve => setTimeout(() => resolve(false), 100)),
+      waitForSocket(this._port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${this._port} is used, make sure that nothing is running on the port`);
+  }
+
+  private async _startWebServer(): Promise<void> {
+    let collectPortResolve = (port: number) => { };
+    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
+    function collectPort(data: Buffer) {
+      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
+      if (regExp)
+        collectPortResolve(parseInt(regExp[1], 10));
+    }
+
+    this._process = exec(this._config.command, {
+      env: {
+        ...DEFAULT_ENVIRONMENT_VARIABLES,
+        ...process.env,
+        ...this._config.env,
+      },
+    });
+    this._process.stdout.pipe(newProcessLogPrefixer()).pipe(process.stdout);
+    this._process.stderr.pipe(newProcessLogPrefixer()).pipe(process.stderr);
+    if (!this._port)
+      this._process.stdout.on('data', collectPort);
+    let processExitedWithNonZeroStatusCodeCallback = (error: Error) => { };
+    this._processExitedWithNonZeroStatusCode = new Promise((_, reject) => processExitedWithNonZeroStatusCodeCallback = reject);
+    this._process.on('exit', code => processExitedWithNonZeroStatusCodeCallback(new Error(`WebServer was not able to start. Exit code: ${code}`)));
+    if (!this._port) {
+      this._port = await collectPortPromise;
+      console.log(`[Playwright Test] Port ${this._port} by process '${this._config.command}' was automatically detected. Starting the tests...`);
+    }
+  }
+
+  private async _waitForAvailability() {
+    assert(this._port);
+    const launchTimeout = this._config.timeout || 60 * 1000;
+    const cancellationToken = { canceled: false };
+    const { timedOut } = (await Promise.race([
+      raceAgainstDeadline(waitForSocket(this._port, 100, cancellationToken), launchTimeout + monotonicTime()),
+      this._processExitedWithNonZeroStatusCode,
+    ]));
+    cancellationToken.canceled = true;
+    if (timedOut) {
+      this.kill();
+      throw new Error(`failed to start web server on port ${this._port} via "${this._config.command}"`);
+    }
+  }
+  public async kill() {
+    assert(this._process);
+    await new Promise(resolve => kill(this._process!.pid, 'SIGTERM', resolve));
```

Comment:
Btw, the process might have exited at this point, due to any reason. Will this throw? I think `kill` would better not throw.